src/Chapter3.hs

```diff
@@ -480,6 +517,37 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data CultureBuilding = Church | Library deriving (Show)
+
+data Castle = Castle String deriving Show
+data Wall = Wall deriving Show
+data House = House Int deriving Show
+
+data City = City
+    { cityCastle :: Maybe Castle
+    , cityWall :: Maybe Wall
+    , cityCultureBuilding :: Maybe CultureBuilding
+    , cityHouses :: [House]
+    } deriving (Show)
+
+buildCastle :: City -> String -> City
+buildCastle city name =
+    city { cityCastle = Just $ Castle name } 
+
+buildHouse :: City -> Int -> City
+buildHouse city c
+    | c > 4 = city
+    | otherwise = city { cityHouses = (House c) : cityHouses city }
```

Comment:
Interesting approach. It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `buildHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂